text/0000-get-expr-type.md

```diff
@@ -0,0 +1,239 @@
+- Feature Name: `get_expr_type`
+- Start Date: 2019-05-30
+- RFC PR: [rust-lang/rfcs#2706](https://github.com/rust-lang/rfcs/pull/2706)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add the ability to retrieve the concrete type of an arbitrary expression so that
+it may be reused by other code that relies on the expression.
+
+# Motivation
+[motivation]: #motivation
+
+> "Macros will become more powerful than you can possibly imagine."
+>
+> – Obi-Wan Kenobi (_supposedly_)
+
+Within the context of a macro, this feature would be very useful.
+
+It would allow for:
+
+- Defining a function that takes the concrete type of a given expression.
+- Calling upon type-level functions based upon a given expression.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Given some value `x`, we can retrieve its type with just `x.type`. This type can
+then be used within various contexts:
+
+```rust
+let x = "hellooo";
+let y: x.type = x; // C equivalent: typeof(x) y = x;
+
+type X = x.type;
+
+assert_eq_type!(X, &str); // taken from `static_assertions`
+
+assert!(<x.type>::default().is_empty());
+```
+
+Note that if the expression resolves to a long operation, the operation will
+_not_ be evaluated.
+
+```rust
+type T = do_stuff().do_more().type;
+```
+
+This feature is especially useful within macros when wanting to generate
+functions or types based off of the given expression.
+
+```rust
+let x: Value = /* ... */;
+
+macro_rules! do_stuff {
+    ($x:expr) => {
+        let y = <$x.type>::new(/* ... */);
+        $x.do_stuff_with(y);
+    }
+}
+
+do_stuff!(x);
+```
+
+When we get `x.type` here, it is no different than just substituting it directly
+with `Value`. This allows for accessing all type-specific functionality.
+
+This isn't possible with the current mechanism for getting the type of an
+expression:
+
+```rust
+macro_rules! do_stuff {
+    ($x:expr) => {
+        fn with_type_of<T>(x: T) {
+            let y = T::new(/* ... */);
+            x.do_stuff_with(y);
+        }
+        with_type_of($x);
+    }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+I am uncertain of how this would be implemented within the compiler but I
+imagine that it would leverage the work that's already been done with `const`
+generics.
+
+Some cases to be aware of are non-simple expressions, such as `1 + 1`. It is
```

Comment:
How does this interact with type inference? Currently, as I understand it, Rust allows the following code:

```rust
let mut v = Vec::new();  //1, 3
v.push(1u8); //2
```

Where type inference works in the following way:

1. `v` has type `Vec<?>`
2. `Vec::push()` is called with a `u8`. This parameter is of type `T` therefore `T` has type `u8`
3. Therefore `v` has type `Vec<u8>`.

Now consider the following code:

```rust
let mut v1 = Vec::new();
type V = v1.type; //What is the type `V` here? 
let mut v2 = V::new();

v2.push(1u16); //Does this line mean `v1` is of type `Vec<u16>`?
v1.push(1u8); //Does this line compile?
```

Would you expect this code to be valid? If not, what is the type of `v1`? Is it `Vec<u8>` because of `v1.push(1u8)` or `Vec<u16>` because `type V = v1.type` and `v2.push(1u16)`?

Reply:
Simplifying a bit:

```rust
let mut v1 = Vec::new();
let mut v2 = <v1.type>::new();
```

What should happen with `v2` is that the known type of `v1` at that point is used. Specifically, we know that `v1: Vec<?T0>` and so this type, *including the names of unsolved inference variables*, is also assigned to `v2`. This would use the same inference variable `?T0` for both `v1` and `v2`. Therefore, when we do `v2.push(1u16)` we should equate `?T0 = u16` and therefore it follows that `v1: Vec<u16>` wherefore `v1.push(1u18);` cannot compile.

Reply:
Seems logical but potentially surprising to users that `v1`'s type was inferred because of how `v2` was used.

Reply:
That's type inference for you... often surprising ;) E.g. we have bidirectional type-checking but folks often don't notice the difference.

We could do some alternative schemes:

2. Generate fresh variables in `expr.type` whenever for each variable.
   Seems less useful?

3. Error when unsolved variables are found in `expr.type`.
   This entails that the full type must be known in `expr`.

Reply:
This isn't surprising at all.  There are already many, many ways to crate a situation in rust where the type of some local `v1` is decided based on how another local `v2` is used.